### PR TITLE
[BugFix][Model] Fix the GLM-5.3-Flash NoPE and KDA convolution paths on Ascend

### PR DESCRIPTION
**Pull request by yiminghub2024**
[BugFix][Model] Fix the GLM-5.3-Flash NoPE and KDA convolution paths on Ascend

### What this PR does / why we need it?

Four things GLM-5.3-Flash needs before it runs correctly on Ascend, and one bug
that reaches well past it.

GLM-5.3-Flash applies no rope, so two paths that assume one had to be taught to
step aside: the QKNorm/rope fusion pass, which fused a rope that is not there,
and the MLA rope-cache operators, which a NoPE layer must not call.

Its KDA layers convolve through `causal_conv1d_update`. The per-request PyTorch
path read `query_start_loc` with `.item()` once per request, which is rejected
outright while an ACL graph is being captured, so decode-FULL capture aborted;
it also looped in Python once per KDA layer on every decode step. Route to the
fused `npu_causal_conv1d_custom` operator where the hardware registers it, and
to a batched torch expression where it does not -- A5 withholds
`RUNTIME_CUSTOM_OPS` (#7157), so custom ops are off there entirely.

The last commit is the one worth reading. **Both** torch paths read
`num_accepted_tokens` as how many of this step's tokens to convolve. It
describes the step *before*: which drafts the sampler keeps is only known once
the model has run. A steady decode accepts one token, so seven of every eight
draft-verify rows fell out and reached the recurrent layers as raw projections.
Their logits could not match a draft, acceptance sat at zero past the first
position, and generation came out garbled.

Upstream's kernel uses the count for one thing, the column a request's history
starts at:

```python
conv_state_token_offset = tl.load(num_accepted_tokens_ptr + idx_seq) - 1
```

and widens the state it writes back to `width - 1 + (seqlen - 1)`, which is the
room that rewind reads from. Both paths now do the same: convolve every row,
take the count as the read offset, and store the history shifted by one, so
that the window ending at this step's a-th token sits at column a - 1.

The per-request path in `vllm_ascend/ops/causal_conv1d.py` is the fallback for
every model whose hardware withholds the fused operator, Qwen3-Next GDN and
Kimi KDA among them, and it had no rewind at all: it only ever read and wrote
the first `width - 1` columns of a row that is allocated `num_spec` wider.
Speculative decoding on those models is affected the same way.

### Does this PR introduce _any_ user-facing change?

No new options or configuration. Speculative decoding on models that fall back
to the torch conv1d path now produces correct output and a usable acceptance
rate where it previously produced neither.

### How was this patch tested?

Unit tests, 35 across three files:

```bash
pytest -sv tests/ut/models/test_glm5next_causal_conv1d.py \
           tests/ut/attention/test_mla_nope_cache_paths.py \
           tests/ut/compilation/test_qknorm_rope_fusion_pass.py
```

The conv1d tests previously checked the batched expression only against the
per-request one, which is what let a mistake shared by both go unseen. They now
also check against the operator's definition written out directly: that every
row of a verify step is convolved, and that reading the stored row at offset
`a - 1` lands on the window ending with this step's a-th token, for every `a`.

On hardware, GLM-5.3-Flash TP8 on Ascend A5 with DFlash2 and seven draft
tokens, per-position acceptance:

- before: `0.000` at every position
- after: `0.583  0.403  0.278  0.181  0.167  0.111  0.069`

Output went from garbled to coherent on the same prompt.

- vLLM main: https://github.com/vllm-project/vllm/commit/b2f685834a6456197e7033966fdef52a23f1abcd

### CODE REVIEW

File: vllm_ascend/models/glm5next/ops/causal_conv1d.py [hunk 1]
```diff
@@ -4,69 +4,249 @@
...
 
-    for key in _UPDATE_FALLBACK_UNSUPPORTED_KWARGS:
-        kwargs.pop(key, None)
-    return _torch_causal_conv1d_update(*args, **kwargs)
+    # Positions past the accepted prefix keep the projection they came in with,
+    # matching the reference. Dropped slots are aimed at a scratch row that is
+    # sliced off, so their duplicate indices cannot clobber a real token.
+    result = torch.cat([x, x.new_zeros(1, dim)], dim=0)
+    destinations = torch.where(keep, token_ids, torch.full_like(token_ids, num_tokens))
+    result.index_copy_(0, destinations.reshape(-1), conv_out.reshape(-1, dim).to(x.dtype))
```
[the remark this answers is not in the document]
Comment (yiminghub2024) [reply vllm_ascend/models/glm5next/ops/causal_conv1d.py]: fixed

File: vllm_ascend/models/glm5next/ops/causal_conv1d.py [hunk 2]
```diff
@@ -4,69 +4,249 @@
...
-        return _npu_triton_conv1d_update(*args, **kwargs)
+    # Store the history shifted by one, so that reading at offset a - 1 next
+    # step lands on the window ending with this step's a-th token. A plain
+    # decode step brings one token, where this is the usual slide by one.
+    write_len = min(state_width, state_len + max_query_len - 1)
+    advanced = history[:, 1 : 1 + write_len].to(states.dtype)
+    # A request with no tokens keeps its row as it was, which the slots the
+    # metadata builder pointed at the null block rely on.
+    held = states.index_select(0, state_slots)[:, :write_len]
+    states[:, :write_len].index_copy_(0, state_slots, torch.where(keep[:, :1].unsqueeze(-1), advanced, held))
```
[the remark this answers is not in the document]
Comment (yiminghub2024) [reply vllm_ascend/models/glm5next/ops/causal_conv1d.py]: fixed

File: vllm_ascend/ops/causal_conv1d.py [hunk 3]
```diff
@@ -229,19 +229,40 @@ def _select_state(i: int) -> torch.Tensor | None:
...
+        A draft-verify step is handed the count the step before it accepted, and
+        that is what the state has to be rewound to: the row carries one window
+        per count that step could have produced, and the one for a tokens sits
+        a - 1 columns in. Mirrors conv_state_token_offset in the upstream
+        kernel. It says nothing about how many tokens this step may convolve --
+        which drafts the sampler keeps is only known once the model has run.
+        """
+        if num_accepted_tokens is None:
+            return 0
+        return max(int(num_accepted_tokens[i].item()) - 1, 0)
```
Comment (nu11ptr7) [inline vllm_ascend/ops/causal_conv1d.py]: With the Triton binding removed, KDA layers now run this PyTorch fallback, and `_read_at` calls `.item()` per request per forward (host sync) - the deleted comment explicitly warned this stalls ACL graph capture at decode-FULL. Worth keeping a sync-free path (or the Triton kernel) for graph mode.

Comment (yiminghub2024) [reply vllm_ascend/ops/causal_conv1d.py]: Thanks for the catch — the .item() reads in _read_at / the per-request loop are indeed host syncs and would abort decode-FULL ACL graph capture.

That path is not what GLM-5.3-Flash KDA decode runs, though. glm5next/kda.py goes through vllm_ascend/models/glm5next/ops/causal_conv1d.py: fused npu_causal_conv1d_custom when the op is registered, otherwise the batched torch update (no .item(), no per-request Python loop). Qwen3-Next GDN and Kimi KDA call the fused op directly as well; they do not fall through this PyTorch implementation when custom ops are withheld.

vllm_ascend/ops/causal_conv1d.py is still used as the per-request reference for the batched-path tests, as the patch_triton rebind of upstream mamba causal_conv1d_update, and on the prefill causal_conv1d_fn (which is not graph-captured). Restoring the Triton kernel is not a good fit here — Ascend Triton still lacks gdc_wait, and the update has to keep the rewind semantics (num_accepted_tokens is the previous step’s accept count, not how many tokens this step convolves).

If we later want the leftover mamba rebind to be graph-safe as well, that can be a follow-up: dispatch the query_start_loc branch onto the batched update and keep the per-request loop as a private test reference. I would rather not fold them together in this PR, or the GLM batched vs per-request tests become tautological.

**Comment by yiminghub2024** [comment]
Retriggered CI with an empty commit (`8743a2b`).

The previous E2E run (https://github.com/vllm-project/vllm-ascend/actions/runs/34471051271) assigned runners for the three `prepare-csrc-cache` jobs at 11:40 UTC, then sat on **Initialize containers** for 70+ minutes with no logs (`BlobNotFound`). `/rerun` only covers failed jobs, and I cannot cancel Actions on this repo.

`pr_test.yaml` concurrency is `cancel-in-progress`, so this SHA should cancel the hung run and start a fresh E2E pass.

**Comment by yiminghub2024** [comment]
`a3-4 card-(part 2-2)` failed on `test_deepseek_v4_dsa_pcp_dspark` ([job](https://github.com/vllm-project/vllm-ascend/actions/runs/34479568893/job/102904950447)):

```
Acceptance rate at draft position 2 is 0.4933, below minimum 0.55
(tolerance 0.03, so need >= 0.52)
```

The sibling test in the same file passed. This PR does not touch DeepSeek-V4; the miss is a noisy DSpark acceptance-rate check, not the GLM-5.3 NoPE/KDA change. Remaining selected tests are still running — I will `/rerun` the failed job after this E2E run finishes.

**Comment by yiminghub2024** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Ruiqiu-Zheng** [comment]
I reviewed the GLM-5.3 KDA causal-conv ordinary non-spec prefill portion of this PR with official GLM-5.3-Flash layer0 BF16 q/k/v conv weights. This is component/wrapper evidence only; it is not a full-model, E2E, MTP/decode/speculative, production, all-rank/all-layer, or whole-PR readiness claim.

Evidence summary:

- Official weights: q/k/v layer0 tensors were range-extracted from `zai-org/GLM-5.3-Flash`, each BF16 `[8192,1,4]`. Raw tensor hashes: q `714fbd54ce53f7ef926b4c26bb67b22bee38758862625a46f4af0db05f9f4381`, k `81a666b5a7f4a6211101de40cab8dd69841d967143e0426fa7fcd1d092a4f528`, v `c3d2ce7063c7bfddc66bbcd4707ab0e1884f5c48a863534f8ebd10fa6496746c`.
- Correctness: using current-main-compatible source plus the focused #15885 causal-conv subset at head `8743a2b5e4be853d12d7e141e32f6101394959f5`, the harness imported the actual patched wrapper and actual `_pack_conv_weight`. Rank0 and rank7 over A/B/C/D deterministic cases were output/state bitwise exact, max abs `0.0`, with no NaN/Inf.
- Performance: actual-wrapper timing on server15 910B2, physical `/dev/davinci6`, included the wrapper guard, transposes, output allocation, native dispatch, and fallback reference path. With 5 warmups and 20 steady observations per arm, rank0 B/D speedups were `4.244x` and `10.389x`; rank7 B/D speedups were `4.276x` and `10.753x`. A/C were also strong.

Coordination note: PR #16251 also touches the ordinary non-spec GLM KDA causal-conv path and calls the same custom op through a different metadata/staging integration surface. The overlap appears partial rather than cleanly duplicate: #15885 carries fallback/native availability behavior, load-time packing, and explicit pack-order/layout validation, while #16251 routes through GDN metadata and non-contiguous state staging/writeback. I am not recommending which whole PR to merge here; this note is limited to the causal-conv ordinary-prefill evidence for #15885.

Remaining non-scientific/review items observed outside the numerical result include CRLF line endings in the causal-conv-relevant files at the tested PR head, plus normal PR-wide review/merge-state handling.
